crates/file_finder/src/file_finder.rs

```diff
@@ -670,7 +679,11 @@ impl PickerDelegate for FileFinderDelegate {
         if raw_query.is_empty() {
             let project = self.project.read(cx);
             self.latest_search_id = post_inc(&mut self.search_count);
-            self.selected_index.take();
+            if self.first_search {
```

Comment:
I'm not quite sure I understand why do we need this and `first_search` in general.

We create a delegate every time we open the file finder, by calling the `FileFinderDelegate::new` adjusted above.
At the same time we pass the actualized history items list — based on which we bubble up the currently opened item to top of the history matches and now working on selecting bubbled up element's sibling.

`update_matches` is being called on the delegate by the `Picker::new` method with `""` as input, right after we've created the whole thing, so this will trigger on the first moment we open the file finder.
But it might change due to e.g. background file reindexing forced us to do the `Picker::refresh` (which we establish in the same `FileFinderDelegate::new`), and our selections will blink.

Last but not least, tests still pass when I remove this `if` block entirely.

Let's either add tests (and maybe a comment) for this if it's needed, or remove it entirely as now it seems to be noop at best.

Reply:
> `update_matches` is being called on the delegate by the `Picker::new` method with `""` as input, right after we've created the whole thing, so this will trigger on the first moment we open the file finder.

This is the exact reason I've added this `first_search` check. Without it, the unconditional `self.selected_index.take()` just cleared the initial `selected_index` value we've computed in `FileFinderDelegate::new()` and it had no effect.

> Last but not least, tests still pass when I remove this if block entirely.

You're right, they do. But it's becasue you removed `self.selected_index.take()` along with the `if`. I'm not sure if this `.take()` call is really unnecessary here or the test for the behaviour it's intehded to implement is just missing. I didn't want to break some use case I'm not aware of, so I only "removed" it for the first `update_matches()` call which I tested. That's the only reason behind this `if`. So, if `take()` is really unnecessary here, I'll gladly remove it.

Thanks for the review and for the ideas! I'll dig deeper.

Reply:
Please, have a look at the fix. It looks to me that the file finder now behaves as expected.